### PR TITLE
suggested rustsim changes (stub)

### PR DESCRIPTION
**Pull request by cmungall**

### CODE REVIEW

File: tests/test_implementations/test_rustsim_implementation.py [hunk 1]
```diff
@@ -25,24 +30,27 @@ def setUp(self) -> None:
...
-        self.oi = oi
+        self.oi = get_adapter(f"rustsim:sqlite:///{db}")
         self.information_content_scores = {
             "CARO:0000000": 21.05,
             "BFO:0000002": 0.7069,
             "BFO:0000003": 14.89,
         }
         self.compliance_tester = ComplianceTester(self)
+        self.predicates = [IS_A, PART_OF]
+        self._rust_closure_table = relationships_to_closure_table(
```
Comment (cmungall) [inline tests/test_implementations/test_rustsim_implementation.py:41]: `_` indicates private method. No need to do this. The implementation should populate the closure table from the wrapped object. See the original oakx-grape code for an example

File: tests/test_implementations/test_rustsim_implementation.py [hunk 2]
```diff
@@ -25,24 +30,27 @@ def setUp(self) -> None:
...
+    # * Testing python bindings for rustsim package.#################################################
     def test_rustsim_jaccard(self):
         """Tests Rust implementations of Jaccard semantic similarity."""
-        subj_ancs = set(self.oi.ancestors(VACUOLE, predicates=[IS_A, PART_OF]))
-        obj_ancs = set(self.oi.ancestors(ENDOMEMBRANE_SYSTEM, predicates=[IS_A, PART_OF]))
-        jaccard = jaccard_similarity(subj_ancs, obj_ancs)
+        subj_ancs = set(self.oi.ancestors(VACUOLE, predicates=self.predicates))
+        obj_ancs = set(self.oi.ancestors(ENDOMEMBRANE_SYSTEM, predicates=self.predicates))
+
+        jaccard = semantic_jaccard_similarity(
```
Comment (cmungall) [inline tests/test_implementations/test_rustsim_implementation.py:51]: still not testing the interface

File: tests/test_implementations/test_rustsim_implementation.py [hunk 3]
```diff
@@ -1,7 +1,12 @@
 import os
 import unittest
 
-from rustsim import get_intersection, jaccard_similarity, mrca_and_score
+from rustsim import (
```
Comment (cmungall) [inline tests/test_implementations/test_rustsim_implementation.py:4]: no need for this

File: src/oaklib/implementations/rustsim/rustsim_implementation.py [hunk 4]
```diff
@@ -16,7 +21,10 @@
     TermSetPairwiseSimilarity,
 )
 from oaklib.datamodels.vocabulary import OWL_THING
-from oaklib.interfaces.basic_ontology_interface import BasicOntologyInterface
+from oaklib.interfaces.basic_ontology_interface import (
+    RELATIONSHIP,
+    BasicOntologyInterface,
```
Comment (cmungall) [inline src/oaklib/implementations/rustsim/rustsim_implementation.py:26]: ???

File: src/oaklib/implementations/rustsim/rustsim_implementation.py [hunk 5]
```diff
@@ -104,20 +104,11 @@ def multiset_most_recent_common_ancestors(
...
-        #     for s in subjects:
-        #         for o in subjects:
-        #             if asymmetric and s >= o:
-        #                 continue
-        #             pairs.append((s, o))
-        #     for (s, o), lca in nx.all_pairs_lowest_common_ancestor(dg, pairs=pairs):
-        #         yield s, o, lca
-        # else:
-        raise NotImplementedError
+        yield from self.wrapped_adapter.multiset_most_recent_common_ancestors(
```
Comment (cmungall) [inline src/oaklib/implementations/rustsim/rustsim_implementation.py:107]: implement in rust

File: src/oaklib/implementations/rustsim/rustsim_implementation.py [hunk 6]
```diff
@@ -74,24 +86,12 @@ def most_recent_common_ancestors(
...
-        #     n = 0
-        #     for a in common:
-        #         if a not in ancs_of_common:
-        #             yield a
-        #             n += 1
-        #     if n == 0:
-        #         yield OWL_THING
-        # else:
-        raise NotImplementedError
+        yield from self.wrapped_adapter.most_recent_common_ancestors(
```
Comment (cmungall) [inline src/oaklib/implementations/rustsim/rustsim_implementation.py:89]: implement in rust

File: src/oaklib/implementations/rustsim/rustsim_implementation.py (hunk 5 above)
Comment (cmungall) [reply src/oaklib/implementations/rustsim/rustsim_implementation.py:107]: or just don't override and implement later

File: src/oaklib/implementations/rustsim/rustsim_implementation.py [hunk 7]
```diff
@@ -229,14 +220,29 @@ def pairwise_similarity(
...
-            subject_ancestors.add(subject)
-        if object_ancestors is None and isinstance(self, OboGraphInterface):
-            object_ancestors = set(self.ancestors(object, predicates=predicates))
-            object_ancestors.add(object)
-        if subject_ancestors is not None and object_ancestors is not None:
-            sim.jaccard_similarity = jaccard_similarity(subject_ancestors, object_ancestors)
+        if subject == object:
+            sim.jaccard_similarity = 1.0
+        else:
+            if not predicates:
```
Comment (cmungall) [inline src/oaklib/implementations/rustsim/rustsim_implementation.py:226]: this is a pretty expensive block to do in the inner loop - can we not just do this in rust?

Comment (cmungall) [reply src/oaklib/implementations/rustsim/rustsim_implementation.py:226]: or at least cache this

File: src/oaklib/implementations/rustsim/rustsim_implementation.py [hunk 8]
```diff
@@ -382,3 +388,24 @@ def information_content_scores(
             object_closure_predicates=object_closure_predicates,
             use_associations=use_associations,
         )
+
+    def relationships(
```
Comment (cmungall) [inline src/oaklib/implementations/rustsim/rustsim_implementation.py:392]: See https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)

File: src/oaklib/implementations/rustsim/rustsim_implementation.py (hunk 7 above)
Comment (justaddcoffee) [reply src/oaklib/implementations/rustsim/rustsim_implementation.py:226]: Yes, probably should do this in Rust - I made a [PR](https://github.com/Knowledge-Graph-Hub/rustsim/pull/27) to do this

**Comment by cmungall** [comment]
is this branch still active?

**Comment by caufieldjh** [comment]
This is fairly outdated at this point (e.g., references to rustsim) so I will close, but will start a new PR with relevant bits we want to keep

**Pull request closed without merging**